addons/mail/models/discuss/mail_guest.py

```diff
@@ -2,12 +2,61 @@
 
 import pytz
 import uuid
+from functools import wraps
+from inspect import Parameter, signature
 
 from odoo.tools import consteq
 from odoo import _, api, fields, models
+from odoo.http import request
 from odoo.addons.base.models.res_partner import _tz_get
 from odoo.exceptions import UserError
 from odoo.addons.bus.models.bus_presence import AWAY_TIMER, DISCONNECTION_TIMER
+from odoo.addons.bus.websocket import wsrequest
+
+
+def add_guest_to_context(func):
+    """ Decorate a function to extract the guest from the request.
+    The guest is then available on the context of the current
+    request.
+    """
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        req = request or wsrequest
+        token = (
+            kwargs.pop("guest_token", None)
+            or req.httprequest.cookies.get(req.env["mail.guest"]._cookie_name)
+            or req.env.context.get("guest_token", "")
+        )
+        parts = token.split(req.env["mail.guest"]._cookie_separator)
+        guest = req.env.context.get("guest", req.env["mail.guest"])
```

Comment:
I don't understand this line, in which case we would use the decorator and the guest is already in the context?
Also shouldn't this use `_get_guest_from_context` to be safe/consistent?

Reply:
When routes are overridden, the first route to be called will extract the guest from the request (and should do so since the guest token should be popped from the parameters). Afterwards, the routes should take the guest that was extracted from the context since the parameter is not available anymore. I'll update to use `_get_guest_from_context` :+1: